### PR TITLE
cipher: add method to get ciphersuite name

### LINKED ISSUE #143: Get a given ciphersuite as a string
As far as I can tell, it's impossible to translate between a `rustls_supported_ciphersuite` and the actual string value of that ciphersuite. For example if I am in a C program, and rustls has negotiated `TLS_ECDH_RSA_WITH_AES_256_CBC_SHA384` for me, I (or specifically psql) would like to display that to the user but the best I can get right now is a u16, I believe, unless I manually copy all of the constants out from the Rust program.

See also rustls/rustls#822.

### PR DESCRIPTION
**Pull request by kevinburke**
Fixes #143.

### CODE REVIEW

File: src/cipher.rs
Comment (djc) [inline src/cipher.rs:109]: Since there is an implementation of `std::io::Write` for `&mut [u8]` I think we could write format directly into the user-provided buffer with `write_buf.write_fmt(format_args!("{:?}", supported_ciphersuite))`. Do you want to try that? The documentation at https://doc.rust-lang.org/std/io/trait.Write.html#impl-Write-6 seems to indicate this will correctly take care of the situation where the cipher suite name is longer than the provided buffer.

**Review by djc — COMMENTED** [review]
Looks pretty good!

File: src/cipher.rs
Comment (kevinburke) [reply src/cipher.rs:109]: > I think we could format directly into the user-provided buffer with write_buf.write_fmt(format_args!("{:?}", supported_ciphersuite)). Do you want to try that?

Sure. I copied this implementation from `rustls_version` - should I update that one as well?

Comment (djc) [reply src/cipher.rs:109]: Sounds good, but maybe do it in a separate commit?

File: src/cipher.rs
Comment (tgeoghegan) [inline src/cipher.rs:106]: Unless I'm mistaken, this will end up invoking the the `std::fmt::Debug` implementation on [`SupportedCipherSuite`](https://docs.rs/rustls/0.19.1/rustls/struct.SupportedCipherSuite.html), which includes quite a bit more than the name. That seems a little risky, since [`Debug` implementations are generally understood not to be stable](https://doc.rust-lang.org/std/fmt/trait.Debug.html). If this function is exposed just so you can get a string to print out to logs from a C program, then it's fine, but `rustls-ffi` should make it clear that the string obtained through this function should not be parsed by code because its format could change at any time. I think mentioning that in the doccomment should suffice.

Comment (djc) [reply src/cipher.rs:106]: Oh, I missed that! What you're looking for is the `CipherSuite`, not the `SupportedCipherSuite`. The former should be easily accessible from the latter, though.

Comment (tgeoghegan) [reply src/cipher.rs:106]: True, but I think my point about `std::fmt::Debug` being unstable still applies.

Comment (kevinburke) [reply src/cipher.rs:106]: I think that's OK, for the use case that I need it for, which is compatibility with the existing psql implementation - the OpenSSL one shows the cipher you use to negotiate TLS, so I'm hoping to show something similar.

Comment (tgeoghegan) [reply src/cipher.rs:106]: Another nit: can we get the cipher suite name as a `&'static str` and avoid allocating a `String`?

Comment (djc) [reply src/cipher.rs:106]: Not right now, but see https://github.com/rustls/rustls/pull/831.

**Comment by kevinburke** [comment]
I'd like to try to write a test for this change that would let me/us actually verify it does what it says and maybe give us a hook to writing more tests in the future. Roughly, I want to:

- Malloc a buffer in Rust
- Convert it to a `* c_char` and pass to the `get_ciphersuite_name` function
- Verify that the correct data was written to the buffer.

Then do the same with a buffer that is shorter than the data I plan to write into it, and then verify the behavior is what we expect in that case.

**Comment by tgeoghegan** [comment]
Elsewhere, we wrap objects handed off to C in `Box` or `Arc`, which forces them onto the heap (and then requires a call to some `*_free` function). The docs for [`CString`](https://doc.rust-lang.org/std/ffi/struct.CString.html) say it's analogous to Rust `String` (with `&CStr` being analogous to `&str`) so I suspect it is also heap allocated, and your approach should be sound.

**Comment by jsha** [comment]
That approach should be sound, but in general we've avoided using NUL-terminated strings in this library because they are a common source of bugs. Instead we try to give a pointer and length (and also guarantee no internal NULs, for convenience).

> Another nit: can we get the cipher suite name as a &'static str and avoid allocating a String?

I like this idea, and I think it's more than a nit. It affects the API. If we can give the caller a pointer to a &' static str, we don't have to provide a corresponding _free() method, because the str lives for the life of the program. *And* the caller doesn't need to provide a buffer. This is much more convenient for the caller.

Since we don't yet have as_str() offered by the rustls library, I think we should fake it for now with a hardcoded map of ciphersuites to string literals. Like so:

```rust
let name = match supported_ciphersuite {
  TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256 => "TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256",
  ...
};
```

The list of all such ciphersuites is at https://docs.rs/rustls/0.20.0/src/rustls/suites.rs.html#117-135. We could then return that as a rustls_str, similar to how rustls_log_level_str does it.

Glad to see you getting to know Rust better, @kevinburke ! Thanks for the contribution.

**Comment by djc** [comment]
I'm kind of on the fence if it's worth going through the trouble of duplicating the cipher suite names here, since I suspect the as_str() API to land in a few days (and we can likely follow up with a 0.20.1 release soonish).

**Comment by kevinburke** [comment]
I think it makes sense to wait for rustls/rustls#831 so we can get the better interface, but unfortunately at that point we're blocked on #88, so this might take a while to actually land.

**Comment by djc** [comment]
I hope to get #88 to at least compile today or tomorrow (I left it to wait until Rustls 0.20 was released, which finally happened on Sunday), so hopefully it won't be too long.

File: src/cipher.rs
Comment (kevinburke) [reply src/cipher.rs]: I've now made this change

**Comment by kevinburke** [comment]
I've updated this PR to add a test - the previous implementation had an error (it was calling debug on the SupportedCipherSuite and not calling `.suite()`, as a result we were writing the struct details to the buffer in addition to the name. Fortunately the test caught it.

**Comment by kevinburke** [comment]
Note the tests currently fail because I wasn't sure how to handle the case where the input buffer is shorter than the amount of bytes we want to write.

**Comment by kevinburke** [comment]
After some discussion on Slack the conclusion is that this should wait until rustls 0.20.1 has been released.

**Comment by jsha** [comment]
To clarify: I think this should return a `rustls_str`. To make that happen we need one of two things:

 - rustls 0.20.1 released, or
 - write a big match statement mapping from cipher suites to string literals.

I actually favor the latter approach, if you're up for it. I think it's not too complicated. You would take the list of names from https://docs.rs/rustls/0.20.0/src/rustls/suites.rs.html#117-135, e.g. TLS13_AES_256_GCM_SHA384, and transform them (e.g. with sed, or regex in your favorite editor) such that you could do:

```rust
match supported_ciphersuite.suite() {
  CipherSuite::TLS13_AES_128_GCM_SHA256 => "TLS13_AES_128_GCM_SHA256",
  ...
}
```

I would recommend doing a further transform after that to replace "TLS13" on the right hand side with "TLS", to match the [IANA-assigned names](https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml):

```rust
match supported_ciphersuite.suite() {
  CipherSuite::TLS13_AES_128_GCM_SHA256 => "TLS_AES_128_GCM_SHA256",
  ...
}
```

**Comment by djc** [comment]
Why do you favor the latter approach?

**Comment by jsha** [comment]
> Why do you favor the latter approach?

Mainly because I wasn't sure how long rustls 0.20.1 would take and wanted to get this change unblocked. Now that 0.20.1 is out, we can go ahead and do this the right way!

**Comment by kevinburke** [comment]
I'll take another pass at this now

File: Cargo.toml
Comment (kevinburke) [inline Cargo.toml:24]: Note this needs to be reverted before merge.

**Review by jsha — APPROVED** [review]
This looks good, thanks! We'll merge once 0.20.2 is released.

**Comment by jsha** [comment]
0.20.2 is released, this can now be updated.

**Review by jsha — APPROVED** [review]
Thanks!

**Pull request merged**
